### ISSUE TITLE
Replacing .y_range does not work as expected in updating Hbar Figure

### ISSUE THREAD

**Issue by joshusre** [bokeh/bokeh#5316]
Broken code example: https://gist.github.com/joshusre/d0a6bd37d712abc662f73c9ece9e80f0

When replacing the .y_range var with FactorRange(factors=['test_error2', 'another_error_msg2']), no update occurs on button click, but when updating .y_range.factors directly, things work as expected.  See working update func in gist.

**Comment by bryevdv** [comment]
summary: updating `.factors` works, replacing ranges wholesale does not (but should). I believe that this is indicative of a more general issue with adding/replacing new models, so this issue may get closed as a dupe in favor of an issue to fix the general problem.

**Comment by GPistre** [comment]
Any update on this? Is it still broken?
I need to set the x_range **after** creating the figure, and doing something such as : 

```
categories = ['a_category', 'another_category']
fig.x_range = FactorRange(factors=list(categories))
fig.x_range.factors = list(categories)
```

doesn't work either.

**Comment by bryevdv** [comment]
no news but I have scheduled it for the next release cycle.

**Comment by bryevdv** [comment]
Actually, this *might*due to some work in `0.12.5` @joshusre @GPistre can one of you test the code linked at the top and report back?

**Comment by joshusre** [comment]
On my box, the original code is still not updating without @bryevdv 's
modified update bar function.  The issue may still be unresolved.

**Comment by joshusre** [comment]
Still not working as expected in 0.12.5 :(

**Comment by bryevdv** [comment]
Just a hopeful thought, we will look at it explicitly for `0.12.6`

**Comment by bryevdv** [comment]
@joshusre can you possibly update your code to be self contained? I'd like to look at this soon but I need a test case to run and that code assume some database being present.

**Comment by bryevdv** [comment]
ping @joshusre again if possible to provide self-contained test case

**Comment by joshusre** [comment]
https://gist.github.com/joshusre/bbb1a1ee40a452f38759c80cde54623e

yes?

**Comment by bryevdv** [comment]
@joshusre Apologies, I somehow missed the notification for your response. I will try to look at it again soon but it may be too late for `0.12.6` at this point.

**Comment by bryevdv** [comment]
OK, it was a py3 issue, need to call `list` around `items`: 
```
pd.DataFrame(list(exec_data_good["ALL_plot"].items()), ...)
```
will try to investigate now.

**Comment by bryevdv** [comment]
OK can definitely reproduce. However, given that there is a workaround (replace the factors, not the entire range) I am going to push this to the next milestone.

**Comment by steve75382** [comment]
Please fix this as I can't dynamically link and unlink two separate graphs that are linked on the y_range.

**Comment by bryevdv** [comment]
I believe that could be accomplished by using `CustomJS` callbacks on each range that update the `factors` of the other (conditionally, based on a button or checkbox if desired)

**Comment by Charles75382** [comment]
How can this be done with a DataRange1D? I'm not using FactorRange and require a flipped y axis. 

I am gearing everything for Bokeh server. I have three pairs of graphs, each one under its own tab. The graphs consist of 3-4 line plots built form datasets of about 10,000+ points for each line. Three of the six graphs also contain a patch glyph.

I read that Bokeh was good for this sort of thing but I got really bad performance when all the graphs were linked to one y-axis. WebGL sped performance up a bunch but it refuses to plot the line glyphs on one of my logarithmic scale graphs (but will draw the patch glyphs for some reason), and since my axes are reversed, it also has issues with that (the graph goes outside the graph area when panning/zooming). So back to Canvas for me. I suppose I could go back to WebGL and get around the logarithmic graphing issue by making patch glyphs look like lines (I think?) but it still has the bounding issue pictured below.

I had some HoverTool tooltips going as well but all the point glyphs for that really slowed things down so it was taken out. 

Even after this trimming I still had bad performance and found it was due to the linked axes across the 6 graphs. So I linked them in pairs to speed up the performance, but I would like to have them all linked... hence to try and at least offer a toggle or simply hardcode a range refresh from the last viewed tab's range when switching views (it seems that when the graphs are all range-linked they are continuously being redrawn/updated even if they are behind an inactive tab). 

I tried looking into Holoviews and DataShader as well but they seemed more geared for Notebooks than Bokeh server and seems overall unnecessary for my situation. My glyph counts are low even if my datasets are medium to large. If you have any pointers to speed things up I'd greatly appreciate it. So far with these concessions the performance is decent but that's running Bokeh server on my 5GHz desktop, I don't know how it's going to perform yet on a VPS with a few virtual cores. 

Attached is a picture showing what I'm doing and one of the aformentioned issues with WebGL.

![issue](https://user-images.githubusercontent.com/17238375/34674065-1e54679e-f44a-11e7-9f4e-f48bf16a5277.png)

**Comment by jbednar** [comment]
> I tried looking into HoloViews and Datashader as well but they seemed more geared for Notebooks than Bokeh server and seems overall unnecessary for my situation.

I don't think HoloViews makes it easy to have multiple axes, though @philippjfr usually proves me wrong whenever I try to make any blanket statement like that. :-)  HoloViews does make it much simpler to use Datashader, which would be essential if you had 400,00 points, but for 40,000 points total Bokeh should usually be pretty reasonably performant.   

All that said, HoloViews and HoloViews+Datashader do have solid support for Bokeh Server applications.  Our examples are overwhelmingly geared towards notebooks because of the ease of converting notebooks into documentation and the great pain and manual labor involved in converting server apps into documentation, which gives a misleading impression.  So please don't be shy about using them in server apps, and please give feedback about any aspects of the documentation that lead you to believe that server apps are less well supported than notebook usage.

**Comment by philippjfr** [comment]
>I don't think HoloViews makes it easy to have multiple axes, though @philippjfr usually proves me wrong whenever I try to make any blanket statement like that. :-)

In this instance you are correct, HoloViews does not allow changing an axis after initialization, but once this issue is resolved I'd like to change that.

**Comment by bryevdv** [comment]
> How can this be done with a DataRange1D? I'm not using FactorRange and require a flipped y axis.
 
Every earlier discussion this issue involved `FactorRange`, and you didn't state anything otherwise. If you are having an issue with `DataRange1d` that is new information. It may be the same issue, or it may be something else. The most helpful thing (a always) would be an  minimum reproducing example code to investigate. 

re: WebGL it needs a lot love. The original maintainer no longer contributes to the project, and none of the remaining core devs have webGL experience. Combine that with the existing backlog of other issues and it's probably going to be some time before any serious webGL re-work is done. That said, it's possible you could pre-transform your data and use a custom tick formatter to fake a log axis with webGL. However this not the appropriate place to discuss those possibilities, or any other possible optimization of usage, the mailing list would be.

**Comment by Charles75382** [comment]
Getting WebGL running well or wrapping some other kind of hardware acceleration API would be huge... hope it can be addressed sooner than later as Canvas can be a bit slow. I know that's no easy undertaking though... It seems also that WebGL doesn't get the memo with the "above" location of the axes locations if you can see from my picture above. If you could clue me in on how to fix that I would appreciate it. Then I could find a workaround for the log axis problem and have good performance all around.

I don't really know what to do as far as optimization goes, doesn't seem to be much that I can do there if LOD will not do anything for line glyphs since they are vectorized into a single glpyh as per the doc. If there is I would appreciate some info on that from the documentation on pydata.org. 

I also don't know why graphs behind an inactive tab widget are still being rendered (or is the linking of the axes causing them to be rendered in the background?) That is a basic optimization that could probably make my whole issue moot. 

Anyway you seem to want to usher support Q's as this is turning into towards a mailing list so I will leave this at that and move there... but if you could provide insight into the above issues I would greatly appreciate it

**Comment by bryevdv** [comment]
The WebGL issue is documented here: https://github.com/bokeh/bokeh/issues/6867 I am happy to discuss it more there.

Re: proving help with other issues, I'd be happy to try, but as I said this is not the appropriate place. The mailing list is here: https://groups.google.com/a/continuum.io/forum/#!forum/bokeh

I'm not trying to be gruff BTW, but sprawling issues that diverge into many unrelated topics make it *harder for us to do work*, o f which there is already more than we can do.

**Comment by bryevdv** [comment]
This was fixed at some point. The original MRE is a bit hard to follow now, here is a much more concise one that demonstrates that updating a complete `FactorRange` with an `hbar` works as expected:
```python
from bokeh.layouts import column
from bokeh.models import Button, CustomJS, FactorRange
from bokeh.plotting import figure, curdoc

plot = figure(y_range=("foo", "bar"))
r = plot.hbar(y=["foo", "bar"], right=[10, 20], height=0.6)

def update():
    plot.y_range = FactorRange(factors=["baz", "quux"])
    r.data_source.data = dict(y=["baz", "quux"], right=[100, 30])

button = Button()
button.on_event("button_click", update)

curdoc().add_root(column(plot, button))
```
![ScreenFlow](https://github.com/user-attachments/assets/cfc1dc1a-2f63-4138-9f30-c6a003c6612d)
